Refactor for clarity, add comments.

matthew-brett opened the pull request:
We were finding it hard to follow the algorithm, and this seemed generally undesirable, so we have attempted a refactor for greater clarity.

We've also used the more standard RNG instance pattern for the random number generation.

The tests are failing, but this seems also to be true of `main`.

lucascolley commented:
https://github.com/numfocus/small-development-grant-proposals/pull/57#issuecomment-3385307622 sounds good to me. What is the diff to make the tests pass? https://github.com/15r10nk/inline-snapshot could make that easier to find.

matthew-brett commented:
OK - I've added some tests to assert the proportions match the weights, and updated the test output to the current test output.

I did a little more refactoring of the project random selection for clarity.

I've also added a `test.yml` file as `.github/tests.pending`, because I can't push with the workflow enabled:

```
 ! [remote rejected] refactor-algorithm -> refactor-algorithm (refusing to allow a Personal Access Token to create or update workflow `.github/workflows/test.yml` without `workflow` scope)
error: failed to push some refs to 'https://github.com/matthew-brett/small-development-grant-proposals'
```
